packages/salesforcedx-vscode-core/src/commands/auth/forceAuthWebLogin.ts

```diff
@@ -22,39 +23,126 @@ import {
   ProgressNotification
 } from '../../notifications/index';
 import { taskViewService } from '../../statuses/index';
+import { telemetryService } from '../../telemetry';
 import { getRootWorkspacePath, isSFDXContainerMode } from '../../util';
 import {
   DemoModePromptGatherer,
   SfdxCommandlet,
   SfdxCommandletExecutor,
   SfdxWorkspaceChecker
 } from '../util';
+import { AuthParams, AuthParamsGatherer } from './authParamsGatherer';
 import { ForceAuthLogoutAll } from './forceAuthLogout';
 
-import { AuthParams, AuthParamsGatherer } from './authParamsGatherer';
+interface DeviceCodeResponse {
+  user_code: string;
+  device_code: string;
+  interval: number;
+  verification_uri: string;
+}
 
-export class ForceAuthWebLoginExecutor extends SfdxCommandletExecutor<
+export class ForceAuthWebLoginContainerExecutor extends SfdxCommandletExecutor<
   AuthParams
 > {
-  protected showChannelOutput = isSFDXContainerMode();
+  protected showChannelOutput = false;
 
   public build(data: AuthParams): Command {
     const command = new SfdxCommandBuilder().withDescription(
       nls.localize('force_auth_web_login_authorize_org_text')
     );
-    if (isSFDXContainerMode()) {
-      command
-        .withArg('force:auth:device:login')
-        .withLogName('force_auth_device_login');
-    } else {
-      command
-        .withArg('force:auth:web:login')
-        .withLogName('force_auth_web_login');
+
+    command
+      .withArg('force:auth:device:login')
+      .withLogName('force_auth_device_login')
+      .withFlag('--setalias', data.alias)
+      .withFlag('--instanceurl', data.loginUrl)
+      .withArg('--setdefaultusername')
+      .withJson();
+
+    return command.build();
+  }
+
+  public execute(response: ContinueResponse<AuthParams>): void {
+    const startTime = process.hrtime();
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: getRootWorkspacePath(),
+      env: { SFDX_JSON_TO_STDOUT: 'true' }
+    }).execute(cancellationToken);
+    let deviceCodeReceived = false;
+
+    channelService.streamCommandStartStop(execution);
+
+    let stdOut = '';
+    execution.stdoutSubject.subscribe(cliResponse => {
+      stdOut += cliResponse.toString();
+
+      if (!deviceCodeReceived) {
+        const authUrl = this.parseAuthUrlFromStdOut(stdOut);
+
+        if (authUrl) {
+          deviceCodeReceived = true;
+          // open the default browser
+          vscode.env.openExternal(vscode.Uri.parse(authUrl, true));
+        }
+      }
```

Comment:
This is the heart of the new functionality.

We will subscribe the to stdOut from the CLI which comes in two parts throughout this process:
1. DeviceCodeResponse when the CLI makes a request to authenticate
2. The OrgAuthInfo response once the user approves the "device" to access their org.

For this PR we only care about the `DeviceCodeResponse`
```js
{
  "user_code": "4VG9UGUD",
  "device_code": "UGxhdGZvcm1DTEk7OC4zNi4yMjYuMjI0OzE2NTExNjI5NzM5NzI7NFZHOVVHVUQ",
  "interval": 5,
  "verification_uri": "https://login.salesforce.com/setup/connect"
}
```
However, there is no guarantee that we will get the complete, parsable JSON in the stdOut stream all at once. So we will attempt to parse the data each time the subscriber is called until we get a complete authUrl from `parseAuthUrlFromStdOut `, then we will stop trying to parse the output from the CLI by flipping the `deviceCodeReceived ` flag.

If there is no authUrl we don't do anything and any errors will be handled by the CLI or SfdxCommandletExecutor as usual. 

For reference, here is the [CLI code](https://github.com/salesforcecli/plugin-auth/blob/main/src/commands/auth/device/login.ts) for this flow and [tests](https://github.com/salesforcecli/plugin-auth/blob/main/test/commands/auth/device/login.test.ts)

Reply:
Thanks for the explanation.  So in `parseAuthUrlFromStdOut` we show an error to the end user if there is a problem parsing the JSON.  Is that the right path is we have an expected flow where we'll get incomplete results?

Reply:
As far as the error message goes, I don't think we necessarily need to write to the channel service but I thought if there is a situation where the auth doesn't work, maybe from the users experience, nothing happens and maybe the VS Code says the command failed in a toast. It might be nice to have something logged to help indicate what went wrong if they report an issue. Unless they know to look in the output channel they wont see this error message, we will mostly just be logging a telemetry event.